### PR TITLE
REST API Basics (spring task)

### PR DESCRIPTION
**Pull request by Peachex**
task - https://github.com/mjc-school/MJC-School/blob/master/stage%20%233/java/module%20%232.%20REST%20API%20Basics/rest_api_basics_task.md

### CODE REVIEW

File: repository/src/test/java/com/epam/esm/dao/creator/SqlGiftCertificateQueryCreatorTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class SqlGiftCertificateQueryCreatorTest {
+    private final SqlQueryCreator queryCreator = new SqlGiftCertificateQueryCreator();
+    private final List<Criteria> criteriaList = new ArrayList<>();
+
+    @Test
+    public void createQueryTest1() {
```
Comment (AliakseiYermachonak) [inline repository/src/test/java/com/epam/esm/dao/creator/SqlGiftCertificateQueryCreatorTest.java:19]: not informative naming for the test

File: repository/src/test/java/com/epam/esm/dao/creator/SqlGiftCertificateQueryCreatorTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @Test
+    public void createQueryTest2() {
+        String expected = "SELECT gift_certificate_id, certificate_name, description, price, duration, create_date," +
+                " last_update_date, tag_id, tag_name FROM gift_certificates RIGHT JOIN gift_certificates_tags ON" +
+                " gift_certificate_id = gift_certificate_id_fk LEFT JOIN tags ON tag_id = tag_id_fk WHERE" +
+                " LOWER(description) LIKE '%value%' ORDER BY create_date asc;";
+        criteriaList.add(new PartMatchSearchCriteria(SqlGiftCertificateColumnName.DESCRIPTION, "value"));
+        criteriaList.add(new FieldSortCriteria(SqlGiftCertificateColumnName.CREATE_DATE, "asc"));
+        String actual = queryCreator.createQuery(criteriaList);
+        assertEquals(expected, actual);
```
Comment (AliakseiYermachonak) [inline repository/src/test/java/com/epam/esm/dao/creator/SqlGiftCertificateQueryCreatorTest.java:38]: What should be the size of the list here?

File: repository/src/test/java/com/epam/esm/dao/mapper/TagMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class TagMapperTest {
+    @Test
+    public void mapRowTest() throws SQLException {
+        List<Tag> expected = new ArrayList<>();
+        expected.add(new Tag("#funny"));
```
Comment (AliakseiYermachonak) [inline repository/src/test/java/com/epam/esm/dao/mapper/TagMapperTest.java:21]: you may use apache CollectionUtils

File: repository/src/test/java/com/epam/esm/dao/creator/SqlGiftCertificateQueryCreatorTest.java (hunk 2 above)
Comment (Peachex) [reply repository/src/test/java/com/epam/esm/dao/creator/SqlGiftCertificateQueryCreatorTest.java:38]: The list must contain two elements in this test.

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        } else {
+            throw new InvalidFieldException("1", "Invalid gift certificate: " + giftCertificate);
+        }
+    }
+
+    @Transactional
+    @Override
+    public boolean delete(String id) {
+        try {
+            Optional<GiftCertificate> giftCertificateOptional = dao.findById(Long.parseLong(id));
```
Comment (AliakseiYermachonak) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:77]: looks more like repository layer's logic

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        } catch (NumberFormatException e) {
+            throw new InvalidFieldException("1", "Invalid tag id (id = " + id + ")");
+        }
+    }
+
+    @Transactional
+    @Override
+    public boolean update(String id, GiftCertificate newGiftCertificate) {
+        try {
+            GiftCertificate oldGiftCertificate = dao.findById(Long.parseLong(id)).orElseThrow(() ->
```
Comment (AliakseiYermachonak) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:96]: same, looks more like repository layer's logic

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @Override
+    public List<GiftCertificate> findAll() {
+        return dao.findAll();
+    }
+
+    @Override
+    public List<GiftCertificate> findCertificatesWithTagsByCriteria(String tagName, String certificateName, String certificateDescription, String sortByName, String sortByDate) {
+        List<Criteria> criteriaList = new ArrayList<>();
```
Comment (AliakseiYermachonak) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:132]: same as above

File: service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        service = new TagServiceImpl(dao);
+    }
+
+    @Test
+    public void insertTest() {
+        Tag tag = new Tag("#new");
+        Mockito.when(dao.insert(tag)).thenReturn(true);
+        Mockito.when(dao.findByName(Mockito.anyString())).thenReturn(Optional.empty());
+        boolean actual = service.insert(tag);
+        assertTrue(actual);
```
Comment (AliakseiYermachonak) [inline service/src/test/java/com/epam/esm/service/impl/TagServiceImplTest.java:37]: verify may also be added

File: service/src/test/java/com/epam/esm/validator/GiftCertificateValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package com.epam.esm.validator;
+
+import org.junit.jupiter.api.Test;
+
+import java.math.BigDecimal;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class GiftCertificateValidatorTest {
```
Comment (AliakseiYermachonak) [inline service/src/test/java/com/epam/esm/validator/GiftCertificateValidatorTest.java:10]: be careful with testing static methods,

**Pull request merged**